worker/docker_client.py

```diff
@@ -0,0 +1,287 @@
+        for filename in os.listdir('/dev'):
+            if filename.startswith('nvidia'):
+                self._nvidia_device_files.append(os.path.join('/dev', filename))
+
+    def _create_connection(self):
+        if self._docker_host:
+            if self._ssl_context:
+                return httplib.HTTPSConnection(self._docker_host,
+                                               context=self._ssl_context)
+            return httplib.HTTPConnection(self._docker_host)
+        return DockerUnixConnection()
+
+    @wrap_exception('Unable to connect to Docker')
+    def test(self):
+        with closing(self._create_connection()) as conn:
+            conn.request('GET', '/containers/json')
+            if conn.getresponse().status != 200:
+                raise DockerException(self._conn.getresponse().read())
+
+    @wrap_exception('Unable to download Docker image')
+    def download_image(self, docker_image, loop_callback):
+        logger.debug('Downloading Docker image %s', docker_image)
+        with closing(self._create_connection()) as conn:
+            conn.request('POST',
+                         '/images/create?fromImage=%s' % docker_image)
+            create_image_response = conn.getresponse()
+            if create_image_response.status != 200:
+                raise DockerException(create_image_response.read())
+
+            # Wait for the download to finish.
+            while True:
+                loop_callback()
+                response = None
+                line = ''
+                while True:
+                    ch = create_image_response.read(1)
+                    if not ch:
+                        break
+                    line += ch
+                    try:
+                        response = json.loads(line)
+                        logger.debug(line.strip())
+                        break
+                    except ValueError:
+                        pass
+                if not response:
+                    break
+                if 'error' in response:
+                    raise DockerException(response['error'])
+
+    @wrap_exception('Unable to start Docker container')
+    def start_container(self, bundle_path, uuid, command, docker_image,
+                        request_network, dependencies):
+        # Set up the command.
+        docker_bundle_path = '/' + uuid
+        docker_command = [
+            'BASHRC=$(pwd)/.bashrc',
+            'cd %s' % docker_bundle_path,
+            # Run as the user that owns the bundle directory. That way
+            # any new files are created as that user and not root.
+            'U_ID=$(stat -c %%u %s)' % docker_bundle_path,
+            'G_ID=$(stat -c %%g %s)' % docker_bundle_path,
+            'sudo -u \\#$U_ID -g \\#$G_ID -n bash -c ' +
+            '"[ -e $BASHRC ] && . $BASHRC; "' +
+            '"export HOME=%s; "' % docker_bundle_path +
+            '\'(%s) >stdout 2>stderr\'' % command.replace('\'', '\'"\'"\''),
```

Comment:
Does this work for any command with any sort of existing quotes?  This always makes me nervous...